### ISSUE TITLE
keras with tensorflow backend is 4x slower than normal keras on GPU machines

### ISSUE THREAD

**Issue by Flomastruk** [tensorflow/tensorflow#38689]
Same model written in tf.keras runs much slower compared to keras imported directly.

System information (whatever a default GPU session in Colab is):
- Linux Ubuntu 16.04
- Tensorflow 2.2, Keras 2.3.1
- CUDA/cuDNN version: 10.1
- GPU model and memory: Tesla K80

Current behavior. Two identical models defined with **tensorflow.keras** and **keras** have similar training time in CPU session (the code for models is identical). However, when GPU is available tensorflow.keras model doesn't get any acceleration whereas keras model performs 4x faster

```
# returns True
tf.test.is_gpu_available(cuda_only=True, min_cuda_compute_capability=None)
```

Expected behavior. Both models must take advantage of GPU acceleration.

The model below attempts to find an embedding with vectors of unit L2 norm. It takes about 2 minutes / epoch to run on CPU. With GPU, tf.keras model runs in about the same time, normal keras model runs in about 25 seconds.

```
import numpy as np
import tensorflow as tf

vocab_size = 5000
sample_size = 300000
embedding_size = 300

data = np.random.randint(0, vocab_size, sample_size)
y_true = np.ones(sample_size, dtype = np.float32)

# Using tensorflow keras backend
from tensorflow.keras.layers import Input, Embedding, Dot, Reshape
from tensorflow.keras.models import Model

# define the model
input_layer = Input((1,))
embed_layer = Embedding(vocab_size
                              , output_dim = embedding_size
                              , embeddings_initializer='normal')
dot_layer = Dot(axes = -1)

embedded = embed_layer(input_layer)
dotted = dot_layer([embedded, embedded])
out = Reshape((1,))(dotted)

model = Model(inputs = [input_layer], outputs = out)
model.compile(optimizer=tf.keras.optimizers.Adam(1e-3), loss = tf.keras.losses.MeanSquaredError())
model.fit([data], y_true, epochs = 1, verbose = True)
```

To switch to normal keras backend replace the two keras import statements with:
```
from keras.layers import Input, Embedding, Dot, Reshape
from keras.models import Model
```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue. tf.keras takes 111s - 122s for one epoch whereas keras takes 25s. Please find the attached gist for [TF v2.1](https://colab.research.google.com/gist/amahendrakar/74a47571985598cab190db7cc82e7ffd/38689-tf.ipynb#scrollTo=aa941XE8pWwk), [TF v2.2.0-rc3](https://colab.research.google.com/gist/amahendrakar/4ac26feaac5520d1f747989bd7f47dc9/38689-tf.ipynb#scrollTo=aa941XE8pWwk), [TF-nightly](https://colab.research.google.com/gist/amahendrakar/3272f5b327157994845add33ae0866b8/38689-tf-nightly.ipynb) and [keras](https://colab.research.google.com/gist/amahendrakar/42eb0fadb346a13115b0bbddefa5963d/38689-keras.ipynb#scrollTo=pcOLKzFkqG4r). Thanks!

**Comment by jvishnuvardhan** [comment]
@Flomastruk I think you need to provide `batch_size` to `model.fit`. Otherwise, by default `model.fit` considers batch_size as 32. When i added `batch_size=128`, your code took around 31 seconds. You could even increase it to 256 (took 15.5 seconds). Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/2c4d061675286525c6aeb06e04f66ce6/untitled104.ipynb). 

In Tensorflow, you can improve the performance even more by placing some ops and data on GPU or CPU.  Please check the resource [here](https://www.tensorflow.org/guide/gpu).Thanks!

Please close the issue if this was resolved for you. Thanks!

**Comment by Flomastruk** [comment]
Thank you, @jvishnuvardhan, this works well

**Comment by Flomastruk** [comment]
Seems like I am experiencing the same issue, even after specifying `batch_size` parameter to `model_fit`. Native keras backend seems to give 2.5x faster training time versus tensorflow.keras backend.

Below is a code for a basic model and simulating random data. With the settings of basic colab GPU session (Tesla P4, rest from specs in the first post of the thread) it takes 16 sec/epoch using tensorflow.keras and 6sec/epoch using native keras backend:
```
## this is fast
from keras.layers import Input, Embedding, Dot, Reshape, Add
from keras.models import Model
from keras.optimizers import Adam
from keras.losses import MeanSquaredError

## this is slow
# from tensorflow.keras.layers import Input, Embedding, Dot, Reshape, Add
# from tensorflow.keras.models import Model
# from tensorflow.keras.optimizers import Adam
# from tensorflow.keras.losses import MeanSquaredError

emb_dim = 200
batch_size = 2**15
vocab_size = 10**4
sample_size = 5*10**6
dataset_key = np.random.choice(vocab_size, size = 2*sample_size).astype(np.int32).reshape(-1,2)
dataset_val = np.random.uniform(size = sample_size).astype(np.float32)

input_target = Input((1,))
input_context = Input((1,))

target_embedding = Embedding(input_dim=vocab_size, output_dim=emb_dim, input_length=1, name='target_embedding')
context_embedding = Embedding(input_dim=vocab_size, output_dim=emb_dim, input_length=1, name='context_embedding')
w_target = target_embedding(input_target)
w_context = context_embedding(input_context)
dot_product = Dot(axes=-1, name='dot_product')([w_target, w_context])
dot_product = Reshape((1,))(dot_product)

model = Model(inputs=[input_target, input_context], outputs=dot_product)
model.compile(loss = MeanSquaredError(), optimizer = Adam(1e-3))
model.fit([dataset_key[:,0], dataset_key[:,1]], dataset_val, batch_size = batch_size, epochs = 5, verbose=1)
```

**Comment by jvishnuvardhan** [comment]
@Flomastruk Can you please check this with `tf-nightly`. I ran your code with `tf.keras` which took 11.4 seconds and `keras` took 23.7 seconds.
Please take a look at the [gist](https://colab.research.google.com/gist/jvishnuvardhan/4558bf318d8fb4b5e0de232d556f1805/untitled192.ipynb) with `tf.keras`. [Here](https://colab.research.google.com/gist/jvishnuvardhan/33383419b48f664af3a9ca7a1ea011ce/untitled193.ipynb) is the gist with `keras`.

```
tf.keras
Time Taken:  11.416040420532227
Keras 
Time Taken:  23.71488118171692
```

Please verify once and close the issue if this was resolved for you. Thanks!

**Comment by Flomastruk** [comment]
@jvishnuvardhan, you are right as usual. Looks like the latest `tf-nightly` performance is actually even faster than `keras`. I get similar times. Thanks again!

**Comment by marko-radojcic** [comment]
I feel that this stage is opportunity to after python installs tensorflow.
Running tensorflow opens replika window and desktop and laptop. That is how
we can solve this problem.

**Comment by marko-radojcic** [comment]
They are analytical and balanced. We just need simple local developer
install

**Comment by marko-radojcic** [comment]
Also just need then mobile authenticator for payment
